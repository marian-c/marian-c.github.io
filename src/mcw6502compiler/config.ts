export type InstructionsInfo = Record<string, { list: string[]; opCodes: Record<string, number> }>;
export type FinalConfig = {
  __isFinal: true;
  // {LDA: {list: ['izx', 'izy'], opCodes: {IZX: 1, IZY: 2}}}
  instructionsInfo: InstructionsInfo;
};

type OptionalProps = never;

export type OptionalConfig = Omit<FinalConfig, OptionalProps | '__isFinal'> &
  Partial<Pick<FinalConfig, OptionalProps>>;

export type Config = OptionalConfig | FinalConfig;

export function makeConfig(initialConfig: Config): FinalConfig {
  if ('__isFinal' in initialConfig) {
    return initialConfig;
  }

  return {
    ...initialConfig,
    instructionsInfo: Object.fromEntries(
      Object.entries(initialConfig.instructionsInfo).map(([ins, modes]) => {
        return [
          ins.toLowerCase(),
          {
            list: modes.list.map((i) => i.toLowerCase()),
            opCodes: Object.fromEntries(
              Object.entries(modes.opCodes).map(([addressingMode, idx]) => {
                return [addressingMode.toLowerCase(), idx];
              }),
            ),
          },
        ];
      }),
    ),
    __isFinal: true,
  };
}
