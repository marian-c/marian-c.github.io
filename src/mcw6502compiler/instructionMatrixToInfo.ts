import type { Instruction } from '@/vendor-in/my-emulator/6502/olc6502_lookup_types';
import type { InstructionsInfo } from '@/mcw6502compiler/config';
import { sureAccess } from '@/helpers/objects';

export function phase2(phase1: Record<string, Record<string, number>>): InstructionsInfo {
  // TODO: error if relative and (absolute or zp0) mode are configured for the same instruction
  //  they are syntactically incompatible, although the 6502 does not fall into this situation,
  return Object.fromEntries(
    Object.entries(phase1).map(([instructionName, config]) => {
      return [instructionName, { opCodes: config, list: Object.keys(config) }];
    }),
  );
}

/**
 * given an instruction matrix, produce the instructionsInfo to be used by the compiler
 */
export const instructionMatrixToInfo = (matrix: Instruction[]): InstructionsInfo => {
  const phase1 = matrix.reduce(
    (acc, el, idx) => {
      const [instructionName, _, addressingModeFn] = el;
      const addresingMode = addressingModeFn.fName;
      if (acc[instructionName] === undefined) {
        acc[instructionName] = {};
      }
      sureAccess(acc, instructionName)[addresingMode] = idx;
      return acc;
    },
    {} as Record<string, Record<string, number>>,
  );

  return phase2(phase1);
};
