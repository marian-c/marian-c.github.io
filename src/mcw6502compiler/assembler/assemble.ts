import { type Config, type FinalConfig, makeConfig } from '@/mcw6502compiler/config';
import {
  type AddressingModeNode,
  DirectiveKind,
  type DirectiveNode,
  InstructionKind,
  type InstructionNode,
  type InstructionWithOperandNode,
  type LabelDefinitionNode,
  NodeKind,
  type ReferenceNode,
  type SymbolDefinitionNode,
  ValueLiteralKind,
  type ValueLiteralNode,
} from '@/mcw6502compiler/parser/ast';
import { sureAccess } from '@/helpers/objects';
import { assertNever } from '@/utils';
import { ErrorShouldNotHappen } from '@/errors/ErrorShouldNotHappen';
import { type GenericResult, GenericResultKind } from '@/types';
import {
  type ErrorResult,
  makeError,
  type ParseResult,
} from '@/mcw6502compiler/parser/parserTypes';

// TODO: defined symbols should be case insensitive
// TODO: validate the values of valuable, for example a given value must make sense for where it's used (zeropage for example)

// TODO: check for invalid references (no use before define, except for labels, all need to exist)

// TODO: check for references to far away

class AssemblerError extends Error {
  constructor(public errorResult: ErrorResult) {
    super(
      `AssembleError encountered ${errorResult.message}, ${errorResult.range}, ${errorResult.lineRange}`,
    );
  }
}

type InternalState = {
  hexList: number[];
  definitions: Map<string, number[]>;
  references: Map<
    number,
    { name: string; len: number; mode: 'as-is' } | { name: string; mode: 'relative' }
  >;
  parseTree: ParseResult;
  startIdx: number | undefined;
};

function numberToBytes(value: number): number[] {
  const result: number[] = [];
  while (true) {
    const r = value & 0x00ff;
    result.push(r);
    value = value >> 8;
    if (value === 0) {
      break;
    }
  }
  return result;
}

function bytesToNumber(bytes: number[]): number {
  return bytes.reduce((acc, curr, idx) => {
    return (curr << (idx * 8)) | acc;
  }, 0);
}

function handleLabelDefinitionNode(
  labelDefinitionNode: LabelDefinitionNode,
  _config: FinalConfig,
  state: InternalState,
) {
  switch (labelDefinitionNode.kind) {
    case NodeKind.labelDefinition:
      const value = state.hexList.length + (state.startIdx || 0);
      const bytes = numberToBytes(value);
      if (bytes.length > 2) {
        throw new ErrorShouldNotHappen(`Calculated label definition location too big ${value}`);
      }
      const toFill = 2 - bytes.length;
      if (toFill !== 0) {
        bytes.push(...new Array(toFill).fill(0));
      }
      state.definitions.set(labelDefinitionNode.value, bytes);
      break;
    default:
      assertNever(labelDefinitionNode.kind);
  }
}

function valueLiteralToNumber(valueLiteral: ValueLiteralNode): number {
  let value: number;
  switch (valueLiteral.valueLiteralKind) {
    case ValueLiteralKind.ascii:
      value = valueLiteral.originalStringValue
        .split('')
        .reverse()
        .reduce((acc, curr, idx) => {
          const v = curr.charCodeAt(0);
          const r = (v << idx) + acc;
          return r;
        }, 0);
      break;
    case ValueLiteralKind.hex:
      value = parseInt(valueLiteral.originalStringValue, 16);
      break;
    case ValueLiteralKind.dec:
      value = parseInt(valueLiteral.originalStringValue, 10);
      break;
    case ValueLiteralKind.bin:
      value = parseInt(valueLiteral.originalStringValue, 2);
      break;
    default:
      assertNever(valueLiteral.valueLiteralKind);
  }

  return value;
}

function valueLiteralToBytes(valueLiteral: ValueLiteralNode): number[] {
  let value: number;
  switch (valueLiteral.valueLiteralKind) {
    case ValueLiteralKind.ascii:
      return valueLiteral.originalStringValue
        .split('')
        .reverse()
        .map((c) => c.charCodeAt(0));
    case ValueLiteralKind.hex:
      value = parseInt(valueLiteral.originalStringValue, 16);
      break;
    case ValueLiteralKind.dec:
      value = parseInt(valueLiteral.originalStringValue, 10);
      break;
    case ValueLiteralKind.bin:
      value = parseInt(valueLiteral.originalStringValue, 2);
      break;
    default:
      assertNever(valueLiteral.valueLiteralKind);
  }

  const result: number[] = [];
  while (true) {
    const r = value & 0x00ff;
    result.push(r);
    value = value >> 8;
    if (value === 0) {
      break;
    }
  }
  return result;
}

function insertValueLiteral(
  limit: 1 | 2,
  valueLiteralNode: ValueLiteralNode,
  _config: FinalConfig,
  state: InternalState,
): void {
  const bytes = valueLiteralToBytes(valueLiteralNode);
  if (bytes.length > limit) {
    throw new AssemblerError(
      makeError(
        `Value literal does not fit in ${limit} byte(s)`,
        valueLiteralNode.range,
        state.parseTree.originalLines,
      ),
    );
  }
  if (limit === 2 && bytes.length === 1) {
    bytes.push(0);
  }
  state.hexList.push(...bytes);
}

function insertBytes(
  limit: number,
  bytes_: number[],
  location: number,
  _config: FinalConfig,
  state: InternalState,
): void {
  const bytes = [...bytes_];
  if (bytes.length > limit) {
    throw new AssemblerError(
      makeError(
        `number of bytes to insert (${bytes.length}) is too big, limit: ${limit}`,
        state.parseTree.program.range,
        state.parseTree.originalLines,
      ),
    );
  }

  const toFill = limit - bytes.length;
  if (toFill !== 0) {
    bytes.push(...new Array(toFill).fill(0));
  }
  state.hexList.splice(location, limit, ...bytes);
}

function handleInstructionWithOperandValueLiteral(
  operandKind: AddressingModeNode['kind'],
  valueLiteralNode: ValueLiteralNode,
  config: FinalConfig,
  state: InternalState,
): void {
  switch (operandKind) {
    case NodeKind.relAddressingMode:
      const nextIdx = state.hexList.length;
      const v = valueLiteralToNumber(valueLiteralNode);
      const r = v - nextIdx - 1;
      // TODO: validate R;
      state.hexList.push(r);
      break;
    case NodeKind.immAddressingMode:
    case NodeKind.zp0AddressingMode:
    case NodeKind.zpxAddressingMode:
    case NodeKind.izxAddressingMode:
    case NodeKind.izyAddressingMode:
    case NodeKind.zpyAddressingMode:
      insertValueLiteral(1, valueLiteralNode, config, state);
      break;
    case NodeKind.absAddressingMode:
    case NodeKind.abxAddressingMode:
    case NodeKind.abyAddressingMode:
    case NodeKind.indAddressingMode:
      insertValueLiteral(2, valueLiteralNode, config, state);
      break;

    default:
      assertNever(operandKind);
  }
}
function handleInstructionWithOperandValueReference(
  operandKind: AddressingModeNode['kind'],
  referenceNode: ReferenceNode,
  _config: FinalConfig,
  state: InternalState,
): void {
  switch (operandKind) {
    case NodeKind.relAddressingMode:
      // push a placeholder
      state.hexList.push(0);
      state.references.set(state.hexList.length - 1, {
        mode: 'relative',
        name: referenceNode.referencedName,
      });
      break;
    case NodeKind.immAddressingMode:
    case NodeKind.zp0AddressingMode:
    case NodeKind.zpxAddressingMode:
    case NodeKind.izxAddressingMode:
    case NodeKind.izyAddressingMode:
    case NodeKind.zpyAddressingMode:
      // push a placeholder
      state.hexList.push(0);
      state.references.set(state.hexList.length - 1, {
        len: 1,
        mode: 'as-is',
        name: referenceNode.referencedName,
      });
      break;
    case NodeKind.absAddressingMode:
    case NodeKind.abxAddressingMode:
    case NodeKind.abyAddressingMode:
    case NodeKind.indAddressingMode:
      // push 2 placeholders
      state.hexList.push(0, 0);
      state.references.set(state.hexList.length - 2, {
        len: 2,
        mode: 'as-is',
        name: referenceNode.referencedName,
      });
      break;

    default:
      assertNever(operandKind);
  }
}

function handleInstructionWithOperand(
  instructionWithOperandNode: InstructionWithOperandNode,
  config: FinalConfig,
  state: InternalState,
): void {
  const operatorConf = sureAccess(
    config.instructionsInfo,
    instructionWithOperandNode.operator.value,
  );

  const operand = instructionWithOperandNode.operand;
  const operandKind = operand.kind;

  switch (operandKind) {
    case NodeKind.absAddressingMode:
      state.hexList.push(sureAccess(operatorConf.opCodes, 'abs'));
      break;
    case NodeKind.immAddressingMode:
      state.hexList.push(sureAccess(operatorConf.opCodes, 'imm'));
      break;
    case NodeKind.zp0AddressingMode:
      state.hexList.push(sureAccess(operatorConf.opCodes, 'zp0'));
      break;
    case NodeKind.abxAddressingMode:
      state.hexList.push(sureAccess(operatorConf.opCodes, 'abx'));
      break;
    case NodeKind.abyAddressingMode:
      state.hexList.push(sureAccess(operatorConf.opCodes, 'aby'));
      break;
    case NodeKind.zpxAddressingMode:
      state.hexList.push(sureAccess(operatorConf.opCodes, 'zpx'));
      break;
    case NodeKind.indAddressingMode:
      state.hexList.push(sureAccess(operatorConf.opCodes, 'ind'));
      break;
    case NodeKind.izxAddressingMode:
      state.hexList.push(sureAccess(operatorConf.opCodes, 'izx'));
      break;
    case NodeKind.izyAddressingMode:
      state.hexList.push(sureAccess(operatorConf.opCodes, 'izy'));
      break;
    case NodeKind.relAddressingMode:
      state.hexList.push(sureAccess(operatorConf.opCodes, 'rel'));
      break;
    case NodeKind.zpyAddressingMode:
      state.hexList.push(sureAccess(operatorConf.opCodes, 'zpy'));
      break;
    default:
      assertNever(operandKind);
  }

  const valuable = operand.value;
  const valuableKind = valuable.kind;

  switch (valuableKind) {
    case NodeKind.valueLiteral:
      handleInstructionWithOperandValueLiteral(operandKind, valuable, config, state);
      break;
    case NodeKind.reference:
      handleInstructionWithOperandValueReference(operandKind, valuable, config, state);
      break;
    default:
      assertNever(valuableKind);
  }
}

function handleInstructionNode(
  instructionNode: InstructionNode,
  config: FinalConfig,
  state: InternalState,
): void {
  const instructionKind = instructionNode.instructionKind;
  switch (instructionKind) {
    case InstructionKind.withoutOperand:
      const operatorConf = sureAccess(config.instructionsInfo, instructionNode.operator.value);
      const operandConf = sureAccess(operatorConf.opCodes, 'imp');
      state.hexList.push(operandConf);
      break;
    case InstructionKind.withOperand:
      handleInstructionWithOperand(instructionNode, config, state);
      break;
    default:
      assertNever(instructionKind);
  }
}

function handleDirectiveNode(
  directiveNode: DirectiveNode,
  config: FinalConfig,
  state: InternalState,
) {
  const lineDirectiveKind = directiveNode.directiveKind;
  switch (lineDirectiveKind) {
    case DirectiveKind.org:
      const address = directiveNode.address;
      // TODO: future: could allow .org directives to reference
      //   symbols if they are resolved by now
      const orgValue = valueLiteralToNumber(address);

      if (state.hexList.length === 0 && state.startIdx === undefined) {
        state.startIdx = orgValue;
      } else {
        const fillLength = orgValue - (state.startIdx || 0) - state.hexList.length;
        if (fillLength < 0) {
          throw new AssemblerError(
            makeError(
              `.org directive referenced an address behind the current PC (diff: ${fillLength})`,
              directiveNode.range,
              state.parseTree.originalLines,
            ),
          );
        }
        state.hexList.push(...new Array(fillLength).fill(0));
      }
      break;
    case DirectiveKind.word:
      const valuable = directiveNode.value;
      switch (valuable.kind) {
        case NodeKind.valueLiteral:
          insertValueLiteral(2, valuable, config, state);
          break;
        case NodeKind.reference:
          // push a placeholder
          state.hexList.push(0, 0);
          state.references.set(state.hexList.length - 2, {
            len: 2,
            mode: 'as-is',
            name: valuable.referencedName,
          });
          break;
        default:
          assertNever(valuable);
      }
      break;
    default:
      assertNever(lineDirectiveKind);
  }
}

function handleSymbolDefinitionNode(
  symbolDefinitionNode: SymbolDefinitionNode,
  _config: FinalConfig,
  state: InternalState,
) {
  const kind = symbolDefinitionNode.kind;
  switch (kind) {
    case NodeKind.symbolDefinition:
      const v = valueLiteralToBytes(symbolDefinitionNode.symbolValue);
      state.definitions.set(symbolDefinitionNode.symbolName, v);
      break;
    default:
      assertNever(kind);
  }
}

function _assemble(
  parseTree: ParseResult,
  initialConfig: Config,
): GenericResult<ErrorResult, { data: Uint8Array; startAddress: number }> {
  const config = makeConfig(initialConfig);

  const state: InternalState = {
    hexList: [],
    definitions: new Map<string, number[]>(),
    references: new Map<
      number,
      { name: string; len: number; mode: 'as-is' } | { name: string; mode: 'relative' }
    >(),
    parseTree,
    startIdx: undefined,
  };

  const program = parseTree.program;

  for (const line of program.lines) {
    if (!line.contents) {
      continue;
    }
    const contentsKind = line.contents.kind;

    switch (contentsKind) {
      case NodeKind.labelDefinition:
        handleLabelDefinitionNode(line.contents, config, state);
        break;
      case NodeKind.instruction:
        handleInstructionNode(line.contents, config, state);
        break;
      case NodeKind.directive:
        handleDirectiveNode(line.contents, config, state);
        break;
      case NodeKind.symbolDefinition:
        handleSymbolDefinitionNode(line.contents, config, state);
        break;
      default:
        assertNever(contentsKind);
    }
  }

  // resolve references
  state.references.forEach((reference, location) => {
    const def = state.definitions.get(reference.name);
    if (def === undefined) {
      // this should not happen as the validation step would cover this situation
      throw new ErrorShouldNotHappen(`Referenced label not defined (${reference.name})`);
    }
    switch (reference.mode) {
      case 'as-is':
        insertBytes(reference.len, def, location, config, state);
        break;
      case 'relative':
        if (def.length > 2) {
          throw new AssemblerError(
            makeError(
              `Relative mode reference (${reference.name}) points to something bigger than 2 bytes ${def}`,
              state.parseTree.program.range,
              state.parseTree.originalLines,
            ),
          );
        }
        // TODO: validate delta
        const v = bytesToNumber(def);
        const delta = v - location - 1;
        state.hexList[location] = delta;
        break;
      default:
        assertNever(reference);
    }
  });

  return {
    kind: GenericResultKind.success,
    result: { data: new Uint8Array(state.hexList), startAddress: state.startIdx || 0 },
  };
}

/**
 * Takes a program, returns hex list
 * program should be complete and validated
 */
export function assemble(
  parseTree: ParseResult,
  initialConfig: Config,
): GenericResult<ErrorResult, { data: Uint8Array; startAddress: number }> {
  try {
    return _assemble(parseTree, initialConfig);
  } catch (e) {
    const isAssemblerError = e instanceof AssemblerError;
    let unexpectedMessage = '[missing]';
    if (e instanceof Error) {
      unexpectedMessage = e.message;
    }
    if (!isAssemblerError) {
      console.error(e);
    }
    return {
      kind: GenericResultKind.error,
      errors: [
        isAssemblerError
          ? e.errorResult
          : makeError(
              `Unexpected assembler error: ${unexpectedMessage}`,
              parseTree.program.range,
              parseTree.originalLines,
            ),
      ],
    };
  }
}
