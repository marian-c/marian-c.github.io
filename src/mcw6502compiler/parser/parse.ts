import { type Config, makeConfig } from '@/mcw6502compiler/config';
import {
  type ErrorResult,
  makeSuccess,
  type ParseResult,
  type Result,
  ResultKind,
} from '@/mcw6502compiler/parser/parserTypes';
import { type LineContentsNode, type LineNode, NodeKind } from '@/mcw6502compiler/parser/ast';
import { parseComment } from '@/mcw6502compiler/parser/parseComment';
import {
  labelDefinitionRegExp,
  parseLabelDefinition,
} from '@/mcw6502compiler/parser/parseLabelDefinition';
import { parseInstruction } from '@/mcw6502compiler/parser/parseInstruction';
import { emptyRegExp, symbolDefinitionLightRegExp } from '@/mcw6502compiler/parser/regExp';
import { parseDirective } from '@/mcw6502compiler/parser/parseDirective';
import { parseSymbolDefinition } from '@/mcw6502compiler/parser/parseSymbolDefinition';
import { expected, testRegExp } from '@/mcw6502compiler/parser/stringInput';

/**
 * Given a string (a source file) and some config, it will return the ParseResult
 */
export function parseSource(originalSource: string, initialConfig: Config): ParseResult {
  const config = makeConfig(initialConfig);

  const sourceNormalized = originalSource
    .replace('\r\n', '\n')
    .replace('\r', '\n')
    .replace('\t', ' ');
  const lines = sourceNormalized.split('\n');
  const meta = { originalLines: lines };
  // TODO: does it work with different line endings?
  let index = 0;
  const lineNodes: LineNode[] = [];
  let errors: ErrorResult[] = [];
  for (const line of lines) {
    const trimmed = line.trim();
    if (trimmed) {
      // is it a comment only line
      if (trimmed.startsWith(';')) {
        const commentResult = parseComment({ index, value: line, meta }, config);
        if (commentResult.kind === ResultKind.Error) {
          errors.push(commentResult);
        } else {
          lineNodes.push({
            kind: NodeKind.line,
            range: commentResult.value.range,
            comment: commentResult.value,
          });
        }
      } else {
        let result: Result<LineContentsNode>;
        // is it a label definition
        if (labelDefinitionRegExp.test(trimmed)) {
          result = parseLabelDefinition({ index, value: line, meta }, config);
        } else if (trimmed.startsWith('.')) {
          result = parseDirective({ index, value: line, meta }, config);
        } else if (symbolDefinitionLightRegExp.test(trimmed)) {
          result = parseSymbolDefinition({ index, value: line, meta }, config);
        } else {
          result = parseInstruction({ index, value: line, meta }, config);
        }
        const lineResultWithComment = (function addComment(
          r: Result<LineContentsNode>,
        ): Result<LineNode> {
          if (r.kind === ResultKind.Error) {
            return r;
          }
          let rest = r.rest;

          let commentResult;
          if (rest.value.trim().startsWith(';')) {
            let comment = parseComment(rest, config);
            if (comment.kind === ResultKind.Error) {
              return comment;
            }
            commentResult = comment;
            rest = commentResult.rest;
          }

          // if there's no comment, it should be empty
          if (!commentResult && !testRegExp(rest, emptyRegExp)) {
            return expected(rest, 'End of input');
          }

          return makeSuccess<LineNode>(
            {
              kind: NodeKind.line,
              range: [r.value.range[0], rest.index],
              contents: r.value,
              comment: commentResult?.value,
            },
            rest,
          );
        })(result);
        if (lineResultWithComment.kind === ResultKind.Error) {
          errors.push(lineResultWithComment);
        } else {
          lineNodes.push(lineResultWithComment.value);
        }
      }
    }

    index += line.length + 1;
  }

  return {
    errors,
    program: {
      kind: NodeKind.program,
      lines: lineNodes,
      range:
        lineNodes.length >= 1
          ? [lineNodes[0]!.range[0], lineNodes[lineNodes.length - 1]!.range[1]]
          : [0, sourceNormalized.length],
    },
    originalLines: lines,
  };
}
