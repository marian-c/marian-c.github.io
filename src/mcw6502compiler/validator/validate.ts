import {
  type ErrorResult,
  makeError,
  type ParseResult,
} from '@/mcw6502compiler/parser/parserTypes';
import { NodeKind } from '@/mcw6502compiler/parser/ast';
import type { Config } from '@/mcw6502compiler/config';

/**
 * if the returned value has length 0, there are no errors
 */
export function validateParseTree(parseTree: ParseResult, _initialConfig: Config): ErrorResult[] {
  const errors: ErrorResult[] = [];
  const identifiers = new Set<string>();

  // check for duplicate identifiers
  for (const line of parseTree.program.lines) {
    let identifier = null;
    if (line.contents?.kind === NodeKind.labelDefinition) {
      identifier = line.contents.value;
    }

    if (line.contents?.kind === NodeKind.symbolDefinition) {
      identifier = line.contents.symbolName;
    }

    // TODO: better error location
    if (identifier !== null) {
      if (identifiers.has(identifier)) {
        errors.push(
          makeError(`Duplicate identifier: ${identifier}`, line.range, parseTree.originalLines),
        );
      }
      identifiers.add(identifier);
    }
  }

  // TODO: validate missing references

  return errors;
}
