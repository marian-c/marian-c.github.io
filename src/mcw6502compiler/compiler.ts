import { parseSource } from '@/mcw6502compiler/parser/parse';
import { type Config, makeConfig } from '@/mcw6502compiler/config';
import { validateParseTree } from '@/mcw6502compiler/validator/validate';
import { assemble } from '@/mcw6502compiler/assembler/assemble';
import type { ErrorResult } from '@/mcw6502compiler/parser/parserTypes';
import type { ExtractSuccessType, GenericResult } from '@/types';
import { GenericResultKind } from '@/types';

export function compile(
  originalSource: string,
  initialConfig: Config,
): GenericResult<ErrorResult, ExtractSuccessType<ReturnType<typeof assemble>>> {
  const config = makeConfig(initialConfig);
  let errors: ErrorResult[] = [];

  const parseResult = parseSource(originalSource, config);
  errors.push(...parseResult.errors);

  if (!errors.length) {
    errors.push(...validateParseTree(parseResult, config));
  }

  if (!errors.length) {
    return assemble(parseResult, config);
  }

  return { kind: GenericResultKind.error, errors };
}
