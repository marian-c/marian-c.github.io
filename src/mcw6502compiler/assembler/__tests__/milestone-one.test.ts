import { instructionMatrixToInfo } from '@/mcw6502compiler/instructionMatrixToInfo';
import { lookup } from '@/vendor-in/my-emulator/6502/olc6502_lookup';
import { parseSource } from '@/mcw6502compiler/parser/parse';
import { validateParseTree } from '@/mcw6502compiler/validator/validate';
import { assemble } from '@/mcw6502compiler/assembler/assemble';
import { assertIsGenericResultSuccess } from '@/utils';

const instructionsInfo = instructionMatrixToInfo(lookup);
const config = { instructionsInfo };

function handleWithError(programBody: string) {
  const programResult = parseSource(programBody, config);
  expect(programResult.errors).toHaveLength(0);

  const validationResult = validateParseTree(programResult, config);
  expect(validationResult).toHaveLength(0);

  return assemble(programResult, config);
}
function handle(programBody: string) {
  const output = handleWithError(programBody);
  assertIsGenericResultSuccess(output);
  return {
    startingAddress: output.result.startAddress,
    data: Array.from(output.result.data).map((n) => n.toString(16).padStart(2, '0')),
  };
}
// TODO: errors when various values are too big

describe('milestone one', () => {
  test('debug zero', () => {
    expect(
      handle(`
  LDX $abcd
  start:
  .word start
    `),
    ).toMatchSnapshot();
  });
  test('debug', () => {
    expect(
      handle(`
  lbl = $5
  a = $a
  ab = $abcd
  LDX $abcd
  start:
    LDX ab
  .word start
  .word lbl
  .word $1234
  .word ab
  .word a
    `),
    ).toMatchSnapshot();
  });
  test('one instruction', () => {
    const r1 = handle(`
  .org $1234
    ldx #$ff
    ldx $abcd;
  .org $1240
    ldx #$0f
    JMP $1245
    JMP $1234
    `);
    const r2 = handle(`
    var = $ff
    var_abs = $abcd
  .org $1234
  start:
    ldx #var
    ldx var_abs
  .org $1240
    ldx #%1111
    JMP end
  end:
    JMP start
    `);

    expect(r1).toMatchInlineSnapshot(`
{
  "data": [
    "a2",
    "ff",
    "ae",
    "cd",
    "ab",
    "00",
    "00",
    "00",
    "00",
    "00",
    "00",
    "00",
    "a2",
    "0f",
    "4c",
    "45",
    "12",
    "4c",
    "34",
    "12",
  ],
  "startingAddress": 4660,
}
`);
    expect(r1).toStrictEqual(r2);
  });
});
test('LCD example (no ascii)', () => {
  const prg = `
PORTB = $6000
PORTA = $6001
DDRB = $6002
DDRA = $6003

E  = %10000000
RW = %01000000
RS = %00100000

  .org $8000

reset:
  ldx #$ff
  txs

  lda #%11111111 ; Set all pins on port B to output
  sta DDRB
  lda #%11100000 ; Set top 3 pins on port A to output
  sta DDRA

  lda #%00111000 ; Set 8-bit mode; 2-line display; 5x8 font
  jsr lcd_instruction
  lda #%00001110 ; Display on; cursor on; blink off
  jsr lcd_instruction
  lda #%00000110 ; Increment and shift cursor; don't shift display
  jsr lcd_instruction
  lda #$00000001 ; Clear display
  jsr lcd_instruction

  ldx #0
print:
  lda message,x
  beq loop
  jsr print_char
  inx
  jmp print

loop:
  jmp loop

message: ;.ascii "Hello, world!"

lcd_wait:
  pha
  lda #%00000000  ; Port B is input
  sta DDRB
lcdbusy:
  lda #RW
  sta PORTA
  lda #RW
  sta PORTA
  lda PORTB
  and #%10000000
  bne lcdbusy

  lda #RW
  sta PORTA
  lda #%11111111  ; Port B is output
  sta DDRB
  pla
  rts

lcd_instruction:
  jsr lcd_wait
  sta PORTB
  lda #0         ; Clear RS/RW/E bits
  sta PORTA
  lda #E         ; Set E bit to send instruction
  sta PORTA
  lda #0         ; Clear RS/RW/E bits
  sta PORTA
  rts

print_char:
  jsr lcd_wait
  sta PORTB
  lda #RS         ; Set RS; Clear RW/E bits
  sta PORTA
  lda #RS   ; Set E bit to send instruction
  sta PORTA
  lda #RS         ; Clear E bits
  sta PORTA
  rts

  .org $80A0
  .word reset
  .word $0000 
  `;
  const r = handle(prg);

  expect(r.startingAddress).toEqual(0x8000);
  const expected = `A2 FF 9A A9 FF 8D 02 60
A9 E0 8D 03 60 A9 38 20
55 80 A9 0E 20 55 80 A9
06 20 55 80 A9 01 20 55
80 A2 00 BD 32 80 F0 07
20 6B 80 E8 4C 23 80 4C
2F 80 48 A9 00 8D 02 60
A9 40 8D 01 60 A9 40 8D
01 60 AD 00 60 29 80 D0
EF A9 40 8D 01 60 A9 FF
8D 02 60 68 60 20 32 80
8D 00 60 A9 00 8D 01 60
A9 80 8D 01 60 A9 00 8D
01 60 60 20 32 80 8D 00
60 A9 20 8D 01 60 A9 20
8D 01 60 A9 20 8D 01 60
60 00 00 00 00 00 00 00
00 00 00 00 00 00 00 00
00 00 00 00 00 00 00 00
00 00 00 00 00 00 00 00
00 80 00 00`
    .split('\n')
    .map((r) => r.split(' '))
    .flat()
    .map((v) => v.toLowerCase());
  expect(r.data).toStrictEqual(expected);
});

describe('errors', () => {
  test('bin too big', () => {
    const r1 = handleWithError(`
      ldx #$FFF
    `);
    expect(r1).toMatchInlineSnapshot(`
{
  "errors": [
    {
      "kind": 1,
      "lineRange": {
        "end": [
          2,
          16,
        ],
        "start": [
          2,
          12,
        ],
      },
      "message": "Value literal does not fit in 1 byte(s)",
      "range": [
        12,
        16,
      ],
    },
  ],
  "kind": 1,
}
`);
  });
});
