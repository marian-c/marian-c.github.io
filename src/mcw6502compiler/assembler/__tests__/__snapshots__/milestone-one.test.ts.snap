// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`milestone one debug 1`] = `
{
  "data": [
    "ae",
    "cd",
    "ab",
    "ae",
    "cd",
    "ab",
    "03",
    "00",
    "05",
    "00",
    "34",
    "12",
    "cd",
    "ab",
    "0a",
    "00",
  ],
  "startingAddress": 0,
}
`;

exports[`milestone one debug zero 1`] = `
{
  "data": [
    "ae",
    "cd",
    "ab",
    "03",
    "00",
  ],
  "startingAddress": 0,
}
`;
